### PR TITLE
DEV: move post flags into database

### PR DESCRIPTION
**Pull request by lis2**
This is preparation for a feature that will allow admins to define their custom flags. Current behaviour should stay untouched.

A basic guardian was also defined. In addition, default translations are handled. Old flags will use translations, but at this point, admin defined flags will not be translated.

### CODE REVIEW

File: app/models/post_action_type.rb [hunk 1]
```diff
@@ -11,51 +11,71 @@ def expire_cache
...
+
     def replace_flag_settings(settings)
-      if settings
-        @flag_settings = settings
+      @flag_settings = settings || FlagSettings.new
+      @all_flags = nil
+    end
+
+    def types
+      if overriden_by_plugin?
```
Comment (jjaffeux) [inline app/models/post_action_type.rb:29]: isnt the correct spelling `overridden` ?

File: app/models/post_action_type.rb [hunk 2]
```diff
@@ -11,51 +11,71 @@ def expire_cache
...
+      @all_flags ||= Flag.all
     end
 
     def auto_action_flag_types
-      flag_settings.auto_action_types
+      if overriden_by_plugin?
+        flag_settings.auto_action_types
+      else
+        flag_enum(all_flags.select { |flag| flag.auto_action_type })
+      end
```
Comment (jjaffeux) [inline app/models/post_action_type.rb:54]: ⛳ 

```suggestion
  return flag_settings.auto_action_types if overriden_by_plugin?
  flag_enum(all_flags.select { |flag| flag.auto_action_type })
```

File: app/models/flag.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  after_destroy :reset_flag_settings!
+
+  default_scope { order(:position) }
+
+  def used?
+    PostAction.exists?(post_action_type_id: self.id) ||
+      ReviewableScore.exists?(reviewable_score_type: self.id)
+  end
+
+  def self.reset_flag_settings!
```
Comment (martin-brennan) [inline app/models/flag.rb:19]: I think on this we could have a brief comment about why `DiscourseEvent` is used (for the multiprocess thing) so people don't think it's weird.

File: app/models/flag.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    DiscourseEvent.trigger(:reload_post_action_types)
+  end
+
+  def system?
+    self.id < 1000
```
Comment (martin-brennan) [inline app/models/flag.rb:24]: might just want to shove `1000` into a `MAX_SYSTEM_FLAG_ID` const in this class

File: app/models/post_action_type.rb [hunk 5]
```diff
@@ -11,51 +11,59 @@ def expire_cache
...
+    end
+
+    def all_flags
+      @all_flags ||= Flag.all
     end
 
     def auto_action_flag_types
-      flag_settings.auto_action_types
+      return flag_settings.auto_action_types if overridden_by_plugin?
+      flag_enum(all_flags.select { |flag| flag.auto_action_type })
```
Comment (martin-brennan) [inline app/models/post_action_type.rb:48]:
```suggestion
      flag_enum(all_flags.select(&:auto_action_type))
```

File: app/models/post_action_type.rb [hunk 6]
```diff
@@ -11,51 +11,59 @@ def expire_cache
...
     end
 
     def public_type_ids
       @public_type_ids ||= public_types.values
     end
 
     def flag_types_without_custom
-      flag_settings.without_custom_types
+      return flag_settings.without_custom_types if overridden_by_plugin?
+      flag_enum(all_flags.select { |flag| !flag.custom_type })
```
Comment (martin-brennan) [inline app/models/post_action_type.rb:61]:
```suggestion
      flag_enum(all_flags.reject(&:custom_type))
```

File: app/models/post_action_type.rb [hunk 7]
```diff
@@ -64,15 +72,25 @@ def notify_flag_type_ids
     end
 
     def notify_flag_types
-      flag_settings.notify_types
+      return flag_settings.notify_types if overridden_by_plugin?
+      flag_enum(all_flags.select { |flag| flag.notify_type })
```
Comment (martin-brennan) [inline app/models/post_action_type.rb:76]:
```suggestion
      flag_enum(all_flags.select(&:notify_type))
```

File: app/models/post_action_type.rb [hunk 8]
```diff
@@ -64,15 +72,25 @@ def notify_flag_type_ids
...
+        flag_settings.topic_flag_types
+      else
+        flag_enum(all_flags.select { |flag| flag.applies_to?("Topic") })
+      end
     end
 
     def custom_types
-      flag_settings.custom_types
+      return flag_settings.custom_types if overridden_by_plugin?
+      flag_enum(all_flags.select { |flag| flag.custom_type })
```
Comment (martin-brennan) [inline app/models/post_action_type.rb:89]:
```suggestion
      flag_enum(all_flags.select(&:custom_type))
```

File: app/serializers/post_action_type_serializer.rb [hunk 9]
```diff
@@ -14,25 +14,25 @@ def is_flag
   end
 
   def name
-    i18n("title")
+    i18n("title", default: object.class.names[object.id])
   end
 
   def description
-    i18n("description", tos_url: tos_url, base_path: Discourse.base_path)
+    i18n("description", vars: { tos_url: tos_url, base_path: Discourse.base_path })
```
Comment (martin-brennan) [inline app/serializers/post_action_type_serializer.rb:21]: FYI you can now do this, like JS objects:

```suggestion
    i18n("description", vars: { tos_url:, base_path: Discourse.base_path })
```

File: db/migrate/20240423054323_create_flags.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+      t.text :description
+      t.boolean :notify_type, default: false, null: false
+      t.boolean :auto_action_type, default: false, null: false
+      t.boolean :custom_type, default: false, null: false
+      t.string :applies_to, array: true, null: false
+      t.integer :position, null: false
+      t.boolean :enabled, default: true, null: false
+      t.timestamps
+    end
+    DB.exec("SELECT setval('flags_id_seq', 1001, FALSE);")
```
Comment (martin-brennan) [inline db/migrate/20240423054323_create_flags.rb:17]: Maybe add a brief comment here to say that all flags with ID < 1000 are for system?

File: lib/plugin/instance.rb [hunk 11]
```diff
@@ -197,6 +197,9 @@ def add_report(name, &block)
 
   # Applies to all sites in a multisite environment. Ignores plugin.enabled?
   def replace_flags(settings: ::FlagSettings.new, score_type_names: [])
+    Discourse.deprecate(
+      "add_to_serializer should not be used to directly override include_*? methods. Use the include_condition keyword argument instead",
```
Comment (martin-brennan) [inline lib/plugin/instance.rb:201]: Oh you added deprecation here, but you didn't update the copy + paste :joy:

File: plugins/chat/assets/javascripts/discourse/lib/chat-message-flag.js [hunk 12]
```diff
@@ -45,9 +48,8 @@ export default class ChatMessageFlag {
     let flagsAvailable = flagModal.site.flagTypes;
 
     flagsAvailable = flagsAvailable.filter((flag) => {
-      return flagModal.args.model.flagModel.availableFlags.includes(
-        flag.name_key
-      );
+      const availableFlags = flagModal.args.model.flagModel.availableFlags;
+      return availableFlags.includes(flag.name_key);
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/lib/chat-message-flag.js:52]: I don't think this needs to be assigned to a variable before checking

File: app/models/post_action_type.rb [hunk 13]
```diff
@@ -11,51 +11,59 @@ def expire_cache
...
-        initialize_flag_settings
-      end
-      @types = nil
+    def initialize_flag_settings
+      @flag_settings = FlagSettings.new
     end
 
-    def ordered
-      order("position asc")
+    def replace_flag_settings(settings)
```
Comment (martin-brennan) [inline app/models/post_action_type.rb:23]: Might just want to make a comment here about how this is deprecated, so people don't use it directly and are aware of the coming plugin deprecation.

**Review by martin-brennan — APPROVED** [review]
Looking good, I mostly commented on cosmetic things, the main issue is that the deprecating warning is incorrect :)

**Pull request merged**